### PR TITLE
route: implement virtual host/route configuration level metadata support

### PR DESCRIPTION
**Pull request by wbpcode**
Commit Message: route: implement virtual host/route configuration level metadata support
Additional Description:
Risk Level: low.
Testing: unit.
Docs Changes: added.
Release Notes: added.
Platform Specific Features: n/a.

### CODE REVIEW

**Comment by wbpcode** [comment]
cc @adisuissa here is the implementation PR. :)

File: docs/root/configuration/advanced/metadata_configurations.rst [hunk 1]
```diff
@@ -0,0 +1,20 @@
+.. _metadata_configurations:
```
Comment (mattklein123) [inline docs/root/configuration/advanced/metadata_configurations.rst:1]: Can you run this through ChatGPT and ask it to edit for conciseness and clarity.

**Review by mattklein123 — CHANGES_REQUESTED** [review]
This makes sense, but beyond the fact that metadata is "arbitrary" how is this going to be used? Is this being added for some control plane use or other debugging? If it's going to be logged by Envoy doesn't this also need to be added to the log/header interpolation code?

/wait

File: docs/root/configuration/advanced/metadata_configurations.rst (hunk 1 above)
Comment (wbpcode) [reply docs/root/configuration/advanced/metadata_configurations.rst:1]: I have let ChatGPT to help me to fix the grammar error but didn't let ChatGPT to revise the entire content.
I will have a try.

**Comment by wbpcode** [comment]
> This makes sense, but beyond the fact that metadata is "arbitrary" how is this going to be used? Is this being added for some control plane use or other debugging? If it's going to be logged by Envoy doesn't this also need to be added to the log/header interpolation code?

This patch just is a supplement/complement to current metadata configurations. AFAIK, there is no any strict rules about how to use `metadata` configurations for now. 🤔  

For example, control plane may use `metadata` of filter chain to distinguishes different chains. log system may use `ROUTE_METADATA` to logging additional attributes of routes or generate per route level stats. In the case of `metadata` of virtual host/route configuration, we mainly use them to store virtual host/route configuration level per wasm/lua filter config (By this way, we needn't change the ABI of wasm/lua filter).

In the future, we may use these `metadata` in other ways, and there is no doubt other users may have their own private usage. So, it's hard to say `how is this going to be used` accurately.

**Review by mattklein123 — APPROVED** [review]

**Comment by wbpcode** [comment]
> But doesn't this have to be exposed to WASM/LUA? I'm just trying to understand how this is going to be used right now. Until it's exposed the only use is by control planes or other systems, right?

Yeah. We do some hacks in our private folk to expose these attributes to wasm/lua and I will create new PR to provide a standard way to expose these new metadata.
(But yesterday, I found a new problem. Although metadata is a good solution to expose route/vh level config to the dynamic filter (wasm or lua based), it couldn't work well with the new `disabled` flag in the #30141. If a wasm filter is markd as disabled by default, the route level config in metadata  won't enable it automatically because only config in the `typed_per_filter_config` will be evaluated.)

Anyway, this PR itself basically simple and harmless, so I will land it first.

**Comment by cetanu** [comment]
Interested in the follow-up PR... is the idea that an individual call is made from lua, for each kind of metadata? or are they merged together somehow in the hierarchy of route config / virtualhost / route

**Comment by wbpcode** [comment]
@cetanu I personally prefer the first one. The users need specify the specific type/level of metadata. But I am open to this.

And in fact this is mainly used by our internal fork. So there maybe no follow-up PR in short term. If you want to do something based on the metadata, issues or PRs are welcome and I will try my best to help. :)

**Pull request merged**